### ISSUE TITLE
rint and nearbyint intrinsic documentation is misleading about exceptions

### ISSUE THREAD

**Issue by RalfJung** [llvm/llvm-project#77561]
The rint intrinsic makes this statement about FP exceptions:

> It may raise an inexact floating-point exception if the operand isn’t an integer.

nearbyint differs from rint by not having that statement.

However, this is misleading, since LLVM also documents "the default LLVM floating-point environment assumes that traps are disabled and status flags are not observable". This means that whether or not rint raises an exception is assumed to be un-observable, and LLVM will freely reorder rint with other operations that might alter FP exception flags. Effectively, rint and nearbyint provide equivalent semantics (as least that's how LLVM frontends should treat the situation), and anyone relying on rint doing anything with exceptions will be disappointed eventually.

It would probably be better to remove any mention of exceptions in the rint and nearbyint docs, and possibly to merge those intrinsics. People relying on exception behavior need to use the constrained intrinsics.

**Comment by RalfJung** [comment]
Which PR closed this issue?

**Comment by dtcxzyw** [comment]
> Which PR closed this issue?

Didn't it resolved by https://github.com/llvm/llvm-project/pull/77191?

**Comment by RalfJung** [comment]
No, that did not change the discussion of exceptions, only of rounding modes. The issue should be reopened.

**Comment by dtcxzyw** [comment]
Oh, sorry for my misreading.

**Comment by arsenm** [comment]
This is an unfortunate leftover from before there was a design for strictfp support. We now have 3 intrinsic that mean the same thing. nearbyint, rint and roundeven all do the same thing. We should probably drop llvm.nearbyint and llvm.rint, and auto upgrade them to roundeven (but this would potentially not be worth the asymmetry with the constrained intrinsics)

**Comment by RalfJung** [comment]
That still seems preferable over the confusion caused when frontends think there is a meaningful difference between these intrinsics.

**Comment by RalfJung** [comment]
That said, when Rust tried to use `roundeven` we [ran into issues](https://github.com/rust-lang/rust/pull/95317#issuecomment-1306215792) since it is less widely supported by platform's libm than the alternatives.

**Comment by arsenm** [comment]
> That said, when Rust tried to use `roundeven` we [ran into issues](https://github.com/rust-lang/rust/pull/95317#issuecomment-1306215792) since it is less widely supported by platforms' libm than the alternatives.

That's a lowering issue. We should just start remapping the calls to whatever call exists instead of expecting exact match like happens now

**Comment by leekillough** [comment]
`rint` and `nearbyint` round a floating-point value to an integer value still expressed in the original FP format, using the current rounding mode in the FP environment.

If there is no instruction which can simultaneously fuse the rounding to integer with the current FP rounding mode followed by a conversion of the FP value to an integer register value, and generate all of the necessary FP exceptions, then the FP rounding to integer and the FP conversion to integer register cannot be fused.

On most architectures, the conversion from FP to integer register rounds towards zero regardless of the current FP rounding mode, while the `rint` and `nearbyint` always use the current rounding mode. Since the `rint` and `nearbyint` round the FP value to an integer expressed as a FP value, the conversion of that FP value to an integer register will not depend on the rounding mode or even whether the FP to int register instruction always rounds a certain way.

If you want round towards 0 and conversion of FP to integer register, and do not care about FP exceptions,  the FP -> int conversions (casts) should be sufficient and should not depend on the current FP rounding mode. If you want the FP value rounded to integer a certain direction other than towards 0, you will need to use `rint` or `nearbyint` first, and then convert the FP value to an integer (register) value.

References:

https://en.cppreference.com/w/c/numeric/math/nearbyint.html

https://en.cppreference.com/w/c/numeric/math/rint.html

https://github.com/tactcomplabs/rev/blob/510ac31ce488554b033e56991005418f2d2d27fd/include/RevInstHelpers.h#L65

https://github.com/tactcomplabs/rev/blob/510ac31ce488554b033e56991005418f2d2d27fd/include/insns/Zfa.h#L21

**Comment by RalfJung** [comment]
> rint and nearbyint round a floating-point value to an integer value still expressed in the original FP format, using the current rounding mode in the FP environment.

LLVM assumes that the current rounding mode of the FP environment is always round-to-nearest (except for strictfp operations). It is UB to execute (non-strictfp) LLVM-generated code in a different rounding mode. Therefore it is nonensical for an LLVM operation to refer to the "current rounding mode", we can juts substitute that with round-to-nearest. At that point the three intrinsics all become equivalent.

**Comment by leekillough** [comment]
If `-ffp-model=strict` is not in effect, I suppose you can make `rint` and `nearbyint` always round-to-nearest-even, but even then, the conversion to an integer register will still require a second instruction, unless the ISA has a FP register -> INT register instruction which allows round-to-nearest-even rounding.

Pseudocode:
```
if FP model is strict:
   Use rint to convert FP value to integer value using current rounding mode
   Use cast to convert FP register to integer register
else
   if a FP register -> integer register instruction exists with round-to-nearest-even:
      Convert the FP register to integer register using round-to-nearest-even rounding
   else
      Use rint to convert FP value to integer value using round-to-nearest-even
      Use cast to convert FP register to integer register
   endif
endif
```
